### ISSUE TITLE
[Question] How would you optimize this?

### ISSUE THREAD

**Issue by levicki** [gorhill/uMatrix#673]
So, let's say I have a set of rules like this:
```
* * * block
* 1st-party cookie allow
* 1st-party css allow
* 1st-party image allow
* 1st-party other allow
* 1st-party script allow
* 1st-party xhr allow
...
// for blogspot.com to work I need this:
blogspot.com blogblog.com image allow
blogspot.com blogger.com css allow
blogspot.com blogger.com frame allow
blogspot.com blogger.com image allow
blogspot.com blogger.com script allow
blogspot.com google.com css allow
blogspot.com google.com frame allow
blogspot.com google.com xhr allow
blogspot.com googleusercontent.com image allow

// and with country redirects being popular also for blogspot.co.uk...
blogspot.co.uk blogblog.com image allow
blogspot.co.uk blogger.com css allow
blogspot.co.uk blogger.com frame allow
blogspot.co.uk blogger.com image allow
blogspot.co.uk blogger.com script allow
blogspot.co.uk google.com css allow
blogspot.co.uk google.com frame allow
blogspot.co.uk google.com xhr allow
blogspot.co.uk googleusercontent.com image allow

// and for blogspot.it...
blogspot.it blogblog.com image allow
blogspot.it blogger.com css allow
blogspot.it blogger.com frame allow
blogspot.it blogger.com image allow
blogspot.it blogger.com script allow
blogspot.it google.com css allow
blogspot.it google.com frame allow
blogspot.it google.com xhr allow
blogspot.it googleusercontent.com image allow

// and for blogspot.de...
blogspot.de blogblog.com image allow
blogspot.de blogger.com css allow
blogspot.de blogger.com frame allow
blogspot.de blogger.com image allow
blogspot.de blogger.com script allow
blogspot.de google.com css allow
blogspot.de google.com frame allow
blogspot.de google.com xhr allow
blogspot.de googleusercontent.com image allow

// and...
```
And I quickly get unmanageable rule list because I can't use regex (or at least some sort of wildcard) to have one set of rules for international domains of the same website. Instead of having to change rules only once for all countries I have to change them all to keep them all working.

@gorhill You said regex is out of the question, any ideas how to handle this more efficiently?

**Comment by CrisBRM** [comment]
He liked a similar idea: https://github.com/gorhill/uMatrix/issues/108

**Comment by levicki** [comment]
Not bad but I would still prefer some sort of wildcard such as * if full regex is out of the question for performance reasons.

In any case there is definitely a need for some feature which will ease list maintenace.

**Comment by gorhill** [comment]
> He liked a similar idea: #108.

There is no regex involved, it's just a request to support a more flexible syntax, everything is still all the same internally: lookup tables.

**Comment by gorhill** [comment]
I don't care to optimize this, I would never use uMatrix this way.

**Comment by ddraws** [comment]
@levicki It's best to use uMatrix dynamically. Unlike uBO, maintaining a rule list is very inefficient in this case.

**Comment by levicki** [comment]
OK, so you are basically telling me "you are holding it wrong"?

How then would you suggest using it to accomplish the filters similar to this?

When I asked "How would you optimize this" I meant the rules for such case.

**Comment by gorhill** [comment]
I am telling you I won't rewrite uMatrix from the ground up, including a yet to define the UI to support regex-based rules. It's someone else's project, not mine.

**Comment by levicki** [comment]
@gorhill 
Can you please point to where did I ask you to rewrite anything, much less from the ground up?

I put the word "question" in the issue subject and I even specifically stated that I am aware you are against the regex and asked for alternatives.

All I wanted is to discuss how to handle issue with TLDs, and see if anyone has any good ideas, no need to get so defensive and rude about a simple question.

**Comment by gorhill** [comment]
> no need to get so defensive and rude about a simple question

The essence of your "simple question" is to keep arguing about supporting regex/wildcard, something you _already_ know I declined, _that_ is rude:

> ... I quickly get unmanageable rule list because I can't use regex (or at least some sort of wildcard) ... You said regex is out of the question

Later:

> ... I would still prefer some sort of wildcard ...

I said it already in the past elsewhere: when I decline a request, please respect that choice.

The bottom line is, the more sites one visits on a regular basis, the more rules will be needed. The matrix UI is there to make managing rules easy. If this is not acceptable, then uMatrix is not what you want for your needs.

**Comment by levicki** [comment]
That is not the same as asking for a full rewrite.

Yes, I did ask whether you would perhaps consider a simple wildcard or if there is any alternative to regex you might consider. You did not answer that at all, you just rage-quit and clossd the issue which was clearly labeled as a question, not even a feature request.

I am sorry if I offended you somehow, I didn't know that simply asking questions and stating own preferences is such a cardinal sin in your book. Thanks for hearing me out and for being so open to attempts at constructive discussion.
